rfc/rfc000-package-server.md

```diff
@@ -1,19 +1,87 @@
+*   A server SHALL have one or more endpoints.
+*   An endpoint SHALL accept a request context and propagate this - allowing access to request-specific values such as the identity of the end user, authorization tokens, and the request's deadline.
+*   An endpoint SHALL respect request cancellation if or when the request's deadline expires.
+*   A server MAY expose information about its endpoints, to allow integration with additional tools.
+
+### Contexts
+
+*   Requests SHALL be executed within a request [context](https://blog.golang.org/context), which the server will pass through the request chain.
+
+### Throttling & Backpressure
+
+*   A server MAY throttle inbound requests and reject requests from clients based on a number of factors.
+*   A server MAY respond with either an Out Of Capacity error, or a Rate Limit Exceeded error when rejecting requests.
+*   A server MAY limit the total number of concurrent requests it can serve.
+*   A server MAY impose rate limits on specific clients.
+*   Rate limit behaviour MAY range from minimum request intervals, to time based, or leaky bucket algorithms.
+*   A server MAY implement a pluggable throttle interface, allowing richer implementations - such as an implementation which shares information across instances of the service.
+
+### SLAs
+
+*   A server MAY report its SLA per endpoint to a discovery system, allowing clients to estimate response time.
+
+### Healthchecks
+
+*   A server SHALL accept registration of healthchecks with a defined interface.
+*   A server MAY register default healthchecks to report the health of built in components of the server.
+*   A server SHALL register an endpoint which can be queried to obtain healthcheck information and status.
+*   A server MAY publish these healthcheck statuses via a pluggable transport.
+
+### Service Discovery
+
+*   A server SHALL register itself with a service discovery mechanism on startup.
+*   A server SHALL attempt to deregister itself with a service discovery mechanism on shutdown.
+*   The discovery mechanism SHALL be interchangeable, and satisfy a defined interface, however the mechanism itself is beyond the scope of this RFC.
+
+### Request Tracing
+
+*   Requests received by the server which are accompanied with tracing information SHALL respect this information and pass this information onto other sub-requests initiated by clients within the server.
+*   The request tracing mechanism SHALL be interchangeable, and satisfy a defined interface, however the mechanism is beyond the scope of this RFC.
+
+### Transport
+
+*   A server SHALL receive and respond to requests via a Transport.
+*   The Transport mechanism SHALL be interchangeable, and satisfy a defined interface, however the mechanism of the transport is beyond the scope of this RFC.
+
+### Codec
+
+*   A server SHALL encode and decode requests and responses via an interchangeable Codec.
```

Comment:
Some very rough proof-of-concept code might be worth pursuing here... I'm concerned that there might be a lot of boilerplate, user-visible casting, or both. Would be nice to see how far we can get without generating code.